crates/pico/src/memo.rs

```diff
@@ -0,0 +1,188 @@
+use pico_core::{
+    container::Container,
+    database::Database,
+    dyn_eq::DynEq,
+    epoch::Epoch,
+    key::Key,
+    node::{Dependency, DerivedNode, DerivedNodeId, NodeKind},
+    params::ParamId,
+    storage::{Storage, StorageMut},
+};
+
+pub enum MemoState {
+    Memoized,
+    Computed,
+}
+
+#[allow(clippy::map_entry)]
+pub fn memo<Db: Database>(
+    db: &mut Db,
+    node_id: DerivedNodeId,
+    inner_fn: fn(&mut Db, ParamId) -> Box<dyn DynEq>,
+) -> MemoState {
+    let mut state = MemoState::Memoized;
+    let current_epoch = db.current_epoch();
+    let time_calculated = if db.storage().derived_nodes().contains_key(&node_id) {
+        if any_dependency_changed(db, &node_id, current_epoch) {
+            let (value, dependencies, time_calculated) =
+                call_inner_fn_and_collect_dependencies(db, node_id.param_id, inner_fn);
+            if let Some(node) = db.storage_mut().derived_nodes().get_mut(&node_id) {
+                if node.value != value {
+                    node.value = value;
+                    state = MemoState::Computed;
+                }
+                node.dependencies = dependencies;
+                node.time_calculated = time_calculated;
+                node.time_verified = current_epoch;
+            } else {
+                db.storage_mut().derived_nodes().insert(
+                    node_id,
+                    DerivedNode {
+                        time_verified: current_epoch,
+                        time_calculated,
+                        dependencies,
+                        inner_fn,
+                        value,
+                    },
+                );
+                state = MemoState::Computed;
+            }
+            time_calculated
+        } else {
+            let node = db
+                .storage_mut()
+                .derived_nodes()
+                .get_mut(&node_id)
+                .expect("node should exist. This is indicative of a bug in Pico.");
+            node.time_verified = current_epoch;
+            node.time_calculated
+        }
+    } else {
+        let (value, dependencies, time_calculated) =
+            call_inner_fn_and_collect_dependencies(db, node_id.param_id, inner_fn);
+        db.storage_mut().derived_nodes().insert(
+            node_id,
+            DerivedNode {
+                time_verified: current_epoch,
+                time_calculated,
+                dependencies,
+                inner_fn,
+                value,
+            },
+        );
+        state = MemoState::Computed;
+        time_calculated
+    };
+    register_dependency(
+        db,
+        NodeKind::Derived(node_id),
+        time_calculated,
+        current_epoch,
+    );
+    state
+}
+
+fn any_dependency_changed<Db: Database>(
+    db: &mut Db,
+    node_id: &DerivedNodeId,
```

Comment:
`DerivedNodeId` implements Copy, we can just accept the param directly. (This is true for a all the param/key/id structs)

Reply:
I'll keep a reference in the container trait though to mimic HashMap interface